### PR TITLE
Transpose optimization

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Optimize in canonical the fusion of consecutive transforms, and removal of transform with identity patterns.

Includes tests. For example:

```mlir
  %0 = "onnx.Transpose"(%arg0)  {perm = [0, 1, 2, 3]} : (tensor<10x11x12x13xf32>) -> tensor<10x11x12x13xf32>
```
is removed and usages of %0 is replaced by usages of %arg0.

Combining transforms this
```mlir
  %0 = "onnx.Transpose"(%arg0)  {perm = [3, 2, 1, 0]} : (tensor<10x11x12x13xf32>) -> tensor<13x12x11x10xf32>
  %1 = "onnx.Transpose"(%0)  {perm = [2, 3, 0, 1]} : (tensor<13x12x11x10xf32>) -> tensor<11x10x13x12xf32>
```
into that
```mlir
    %0 = "onnx.Transpose"(%arg0) {perm = [1, 0, 3, 2]} : (tensor<10x11x12x13xf32>) -> tensor<11x10x13x12xf32>
```

Obviously, both optimizations combine to remove redundant transforms like below.
```mlir
  %0 = "onnx.Transpose"(%arg0)  {perm = [3, 2, 1, 0]} : (tensor<10x11x12x13xf32>) -> tensor<13x12x11x10xf32>
  %1 = "onnx.Transpose"(%0)  {perm = [3, 2, 1, 0]} : (tensor<13x12x11x10xf32>) -> tensor<10x11x12x13xf32>
```

### CODE REVIEW

**Comment by doru1004** [comment]
Can you update the description of the patch with an example of your transformations please?

**Comment by AlexandreEichenberger** [comment]
@doru1004 done.

**Comment by doru1004** [comment]
> @doru1004 done.

Thanks a lot for updating the description.

File: src/Transform/ONNX/Combine.cpp [hunk 1]
```diff
@@ -19,6 +19,43 @@ using namespace mlir;
...
+    resPerm.emplace_back(initialPerm[index]);
+  }
+  // Convert to Array of Attributes.
+  ArrayRef<int64_t> resPermRefs(resPerm);
+  return rewriter.getI64ArrayAttr(resPermRefs);
+}
+
+/// Test if the permute pattern correspond to an identity pattern.
+/// Identity patterns are {0, 1, 2, ... , rank -1}.
+bool IsIdentityPermuteVector(ArrayAttr &permAttr) {
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Transform/ONNX/Combine.cpp:47]: Thanks, created #286 to address your concerns. Thanks for contributing to reviews, much appreciated.

**Pull request merged**